CIPs/0017.md

```diff
@@ -0,0 +1,96 @@
+
+# CIP [0017]: Integrate Plumo Proofs into the lightest sync protocol
+
+- Date: 2020-07-27
+- Author: @lucasege
+- Status: WIP
+
+## Overview
+
+In order to continue supporting the dream of a seamless mobile sync experience, the lightest sync protocol must be extended to support PLUMO Zk-SNARKs which can quickly verify transitions from epoch X to epoch X + Y, where Y could be on the order of 10s-100s of epochs.
+
+The prover/verifier for plumo proofs have already been implemented, and the exposed go interface is available [here](https://github.com/celo-org/celo-bls-go)
+
+## Goals
+
+- Support adding/storing proofs on full nodes
+- Gossip these proofs to propogate them through the network
+- Extend the `les` protocol such that the server will provide these proofs in place of headers (when appropriate), and the client will consume/verify proofs accordingly.
+
+## Proposed Solution
+
+We propose to add two new RPCs: 
+* `AddProof(proof []byte, firstEpoch uint, lastEpoch uint, versionNumber uint)`
+* `Proofs() ([][]byte)`.
+
+`versionNumber` here refers to the snark circuit version, for more info refer to #upgradeability below.
+
+When an eventual "proving service" generates a new proof it will publish this proof to a full node that it is running, which will then propagate the proof throughout the network.
+
+This `AddProof` function will then utilize the `verificationKey` corresponding to the `versionNumber` and the corresponding validator sets for these epochs to verify the proof. If successful, it will then store the proof in a new database `plumoproofdata`, keyed by it's combination of first and last epoch. 
+
+To propagate these proofs throughout the network, we also propose to extend the `eth` protocol with three new message types: 
+```
+NewPlumoProof (0x019)
+[[firstEpoch, lastEpoch, versionNumber], ...]
+Specify one or more new plumo proofs which have appeared on the network.
+This is intended to be honored with a proceeding GetPlumoProofsMsg to request the proof.
+
+GetPlumoProofsMsg (0x1a)
+[[firstEpoch, lastEpoch, versionNumber], ...]
+Sent from node A to node B with the set of proofs that node A currently has, requires node B to return a PlumoProofsMsg.
```

Comment:
In steady state, sending the `GetPlumoProodsMsg` with the list of proofs the node has will be much larger than sending a light of the proofs the node wants. Especially for use by long-running nodes who send `GetPlumoProofs` in response to a `NewPlumoProof` message indicating the existence of a proof the node does not have, it will be more efficient to send a message containing only the identifier for the proof the node wants. Adding a flag, or other indicator in the request to allow for requesting specific proof(s) would be beneficial.

Another case where receiving every proof the requestor does not have may not be ideal is if the requester is only interested in a certain subset of proofs (e.g. proofs with the latest version number, or proofs after a certain epoch)

Reply:
Yes certainly, I was wondering if I should split this message into two messages, one to request specifics and one to request any you may be missing.

I like the idea of including this in one message since the logic is of course very similar. 

I'll adapt the spec to follow this idea ^ thanks